### PR TITLE
Use OpenTermsArchive fetch feature

### PR DESCRIPTION
**Pull request by martinratinaud**
Please, review https://github.com/OpenTermsArchive/contribution-tool/pull/45 first

And in order to test

Clone and launch this url http://localhost:3000/en?destination=OpenTermsArchive/sandbox

### CODE REVIEW

File: src/translations/en/service.json [hunk 1]
```diff
@@ -7,6 +7,7 @@
   "dialog.start.title": "Some points of attention",
   "error.cta": "Let us know!",
   "error.title": "We're sorry, an error occured",
+  "pageNotAccurate": "Page does not seem accurate? Try differently.",
```
Comment (MattiSG) [inline src/translations/en/service.json:10]: If that is a call to action, it's good to mention it in the key 🙂 
```suggestion
  "pageNotAccurateCTA": "Page does not seem accurate? Try differently.",
```

File: src/translations/en/service.json [hunk 2]
```diff
@@ -18,6 +19,8 @@
   "form.label.json": "JSON",
   "form.select": "Select...",
   "form.serviceName": "Name",
+  "form.executeClientScripts": "Headless browser",
+  "form.executeClientScripts.more": "Use headless browser for fetching document",
```
Comment (MattiSG) [inline src/translations/en/service.json:23]:
```suggestion
  "form.executeClientScripts": "Switch JavaScript on",
  "form.executeClientScripts.more": "If the page you try to load is interactive, rendering it with JavaScript enabled might improve rendering.",
```

File: ops/roles/infra/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+---
+- name: Install common required packages
+  apt:
+    pkg:
+      - build-essential
+      - curl
+      - git
+      - zip
```
Comment (MattiSG) [inline ops/roles/infra/tasks/main.yml:8]: Why are those necessary? 🤔

**Review by MattiSG — COMMENTED** [review]
There are too many things I do not understand in TypeScript and in the overall architecture of this application to give a proper review. I would need to pair on it to do so, but I will not have time for that this week.

File: ops/roles/ota/github-clone/tasks/main.yml [hunk 4]
```diff
@@ -19,34 +18,54 @@
...
-  community.docker.docker_prune:
-    containers: yes
-    images: yes
-    networks: false
-    volumes: false
-    builder_cache: false
+- name: Delete previous build
+  file:
+    state: absent
+    path: '/home/{{ ansible_user }}/{{ directory }}/dist'
```
[the remark this answers is not in the document]
Comment (martinratinaud) [reply ops/roles/ota/github-clone/tasks/main.yml:36]: Done

File: ops/roles/ota/github-clone/tasks/main.yml [hunk 5]
```diff
@@ -19,34 +18,54 @@
...
+
+- name: Install npm packages
+  command:
+    cmd: npm install
+    chdir: '/home/{{ ansible_user }}/{{ directory }}'
+  tags:
+    - start
+    - update
+
+- name: Build Open Terms Archive Contribution tool
```
[the remark this answers is not in the document]
Comment (martinratinaud) [reply ops/roles/ota/github-clone/tasks/main.yml:49]: Renamed it to `roles/ota` as in OpenTermsArchive

File: src/modules/OTA-api/services/open-terms-archive.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  await launchHeadlessBrowser();
+  const { content, mimeType } = await fetch({
+    url: documentDeclaration.fetch,
+    executeClientScripts: documentDeclaration.executeClientScripts,
+    cssSelectors: documentDeclaration.select,
+    config,
+  });
+  await stopHeadlessBrowser();
+
+  const mdContent = await filter({
```
[the remark this answers is not in the document]
Comment (martinratinaud) [reply src/modules/OTA-api/services/open-terms-archive.ts:14]: Done

File: src/modules/Scraper/utils/downloader.ts [hunk 7]
```diff
@@ -7,14 +7,18 @@ import {
...
-import puppeteer from 'puppeteer-extra';
-const DOWNLOAD_TIMEOUT = 30 * 1000;
+import type { Page, Browser } from 'puppeteer'; // from open-terms-archive
+import puppeteer from 'puppeteer-extra'; // from open-terms-archive
 const logDebug = debug('ota.org:debug');
+import {
+  getVersion,
+  stopBrowser,
+  launchBrowser,
+} from 'modules/OTA-api/services/open-terms-archive';
```
[the remark this answers is not in the document]
Comment (martinratinaud) [reply src/modules/Scraper/utils/downloader.ts:19]: I moved it to tyhe Common module as I agree this naming is not optimal and was more dedicated to retrieve data from the Open Terms Archive legacy API

File: ops/roles/infra/tasks/main.yml (hunk 3 above)
Comment (martinratinaud) [reply ops/roles/infra/tasks/main.yml:8]: `git` in order to `git clone` the repository

`zip` is not, I remove it

**Review by clementbiron — CHANGES_REQUESTED** [review]

File: ops/roles/infra/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  apt:
+    pkg:
+      - chromium
+    update_cache: yes
+    state: latest
+
+  # See https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md#recommended-enable-user-namespace-cloning
+- name: Enable user namespace cloning to allow running Chromium in a sandbox
+  command: sysctl -w kernel.unprivileged_userns_clone=1
+  when: ansible_facts['architecture'] != "aarch64"
```
Comment (Ndpnt) [inline ops/roles/infra/tasks/main.yml:36]: If you do not have a vagrant file, this is not necessary

File: src/translations/fr/service.json [hunk 9]
```diff
@@ -18,6 +20,8 @@
   "form.label.json": "JSON",
   "form.select": "Sélectionner...",
   "form.serviceName": "Nom",
+  "form.executeClientScripts": "Activer le JavaScript",
```
Comment (Ndpnt) [inline src/translations/fr/service.json:23]: > Executer les scripts de la page

File: src/translations/fr/service.json [hunk 10]
```diff
@@ -18,6 +20,8 @@
...
+  "form.executeClientScripts.more": "Si la page que vous essayez de charger est interactive, le rendre avec JavaScript activé peut améliorer le rendu.",
```
Comment (Ndpnt) [inline src/translations/fr/service.json:24]: > Si la page que vous souhaiter afficher charge son contenu de manière dynamique, executer les scripts peut améliorer le rendu.

File: src/translations/fr/service.json [hunk 11]
```diff
@@ -7,6 +7,8 @@
   "dialog.start.title": "Quelques points d'attention",
   "error.cta": "Faites-nous savoir !",
   "error.title": "Nous sommes désolés, une erreur est survenue",
+  "pageNotAccurate.desc": "Un problème avec le chargement de la page ?",
+  "pageNotAccurate.cta": "Cliquez ici, nous allons essayer autrement.",
```
Comment (Ndpnt) [inline src/translations/fr/service.json:11]: > Essayer une autre méthode de chargement

File: src/translations/en/service.json [hunk 12]
```diff
@@ -18,6 +20,8 @@
   "form.label.json": "JSON",
   "form.select": "Select...",
   "form.serviceName": "Name",
+  "form.executeClientScripts": "Switch JavaScript on",
+  "form.executeClientScripts.more": "If the page you try to load is interactive, rendering it with JavaScript enabled might improve rendering.",
```
Comment (Ndpnt) [inline src/translations/en/service.json:24]: See changes suggested on the french translations

File: src/modules/OTA-api/services/open-terms-archive.ts [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    content,
+    mimeType,
+    documentDeclaration: {
+      location: documentDeclaration.fetch,
+      contentSelectors: documentDeclaration.select,
+      noiseSelectors: documentDeclaration.remove,
```
[the remark this answers is not in the document]
Comment (Ndpnt) [reply src/modules/OTA-api/services/open-terms-archive.ts]: Our API is not really inconsistent, is just that there is a difference between the document declaration and the representation in the code, because the purpose is not the same and each have terms which seems more appropriate to the context.

Our API is not really inconsistent, it's just that there is a difference between the syntax of user document declaration and the document declaration representation in the code, as the purpose is not the same and each has terms that seem more appropriate to their context. However, they should probably be named differently to avoid any ambiguity.

**Comment by Ndpnt** [comment]
> There are too many things I do not understand in TypeScript and in the overall architecture of this application to give a proper review. I would need to pair on it to do so, but I will not have time for that this week.

Same for me. So I just made a light review.

File: src/translations/en/service.json [hunk 14]
```diff
@@ -7,6 +7,8 @@
   "dialog.start.title": "Some points of attention",
   "error.cta": "Let us know!",
   "error.title": "We're sorry, an error occured",
+  "pageNotAccurate.desc": "A problem with the loading of the page?",
```
[the remark this answers is not in the document]
Comment (clementbiron) [reply src/translations/en/service.json:10]: Agree 👍

File: src/modules/Scraper/utils/downloader.ts [hunk 15]
```diff
@@ -48,49 +52,85 @@ const removeBaseTag = async (page: Page) => {
   });
 };
 
+const outputPageLogs = (page: Page) => {
+  page.on('console', (consoleObj: any) => logDebug('>> in page', consoleObj.text()));
```
[the remark this answers is not in the document]
Comment (martinratinaud) [reply src/modules/Scraper/utils/downloader.ts]: Added a check to do this only when not in production as logs can be useful for debugging

File: ops/roles/infra/tasks/main.yml (hunk 8 above)
Comment (martinratinaud) [reply ops/roles/infra/tasks/main.yml:36]: 👍

File: src/translations/en/service.json (hunk 12 above)
Comment (martinratinaud) [reply src/translations/en/service.json:24]: ok done

**Comment by martinratinaud** [comment]
All comments have been implemented and branch has been fixed up and rebased.

One last review is mandatory

Site is deployed on preproduction for test https://contribute.preprod.opentermsarchive.org/en/service?destination=OpenTermsArchive/sandbox&localPath=/Users/martin/Workspace/ambanum/contrib-declarations/declarations&versionsRepo=OpenTermsArchive/contrib-versions&url=https%3A%2F%2Fwww.caroll.com%2Ffr_fr%2Fgeneral%2Fconditions-generales-de-vente.html
@Ndpnt @clementbiron

**Review by clementbiron — COMMENTED** [review]
image
Type select is empty ?!

**Comment by martinratinaud** [comment]
> image Type select is empty ?!

Just fixed this
